server/src/main/java/org/opensearch/tasks/TaskCancellationMonitoringService.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.tasks;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.search.SearchShardTask;
+import org.opensearch.common.component.AbstractLifecycleComponent;
+import org.opensearch.common.metrics.CounterMetric;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+/**
+ * This monitoring service is responsible to track long-running(defined by a threshold) cancelled tasks as part of
+ * node stats.
+ */
+public class TaskCancellationMonitoringService extends AbstractLifecycleComponent
+    implements
+        TaskManager.TaskCancellationListener,
+        TaskResourceTrackingService.TaskCompletionListener {
+
```

Comment:
The interface looks inconsistent. It gives a sense that task cancellation monitoring would only be done for tasks that have resource tracking enabled. Will prefer better strong with TaskManager(as an authoritative source of truth) rather than listeners on some other dependent service

Reply:
I agree with you on this. I used completion listener for resource tracking service as it was the only one available. I have similar listener at task manager level, let me know what you think.